staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -440,3 +443,126 @@ func CloneHeader(in http.Header) http.Header {
 	}
 	return out
 }
+
+var codeMatcher = regexp.MustCompile("^[0-9]{3}$")
+
+type WarningHeader struct {
+	Code  string
+	Agent string
+	Text  string
+}
+
+func ParseWarningHeader(header string) (WarningHeader, error) {
+	parts := strings.SplitN(header, " ", 3)
+	if len(parts) != 3 {
+		return WarningHeader{}, errors.New("invalid warning header: fewer than 3 segments")
+	}
+	code, agent, textAndDate := parts[0], parts[1], parts[2]
+	// verify code format
+	if !codeMatcher.Match([]byte(code)) {
+		return WarningHeader{}, errors.New("invalid warning header: code segment is not 3 digits")
+	}
+	// verify agent presence
+	if len(agent) == 0 {
+		return WarningHeader{}, errors.New("invalid warning header: empty agent segment")
+	}
+	// verify textAndDate presence
+	if len(textAndDate) == 0 {
+		return WarningHeader{}, errors.New("invalid warning header: empty text segment")
+	}
+	// extract text
+	text, _, err := parseQuotedString(textAndDate)
+	if err != nil {
+		return WarningHeader{}, fmt.Errorf("invalid warning header: %v", err)
+	}
+	if len(text) == 0 {
+		return WarningHeader{}, errors.New("invalid warning header: empty text segment")
+	}
+	return WarningHeader{Code: code, Agent: agent, Text: text}, nil
+}
+
+func parseQuotedString(quotedString string) (string, int, error) {
+	read := 0
+
+	if len(quotedString) == 0 {
+		return "", 0, errors.New("invalid quoted string: 0-length")
+	}
+
+	if quotedString[0] != '"' {
+		return "", 0, errors.New("invalid quoted string: missing initial quote")
+	}
+	read++
+	quotedString = quotedString[1:]
+
+	result := &bytes.Buffer{}
+	escaping := false
+	closedQuote := false
+
+	for i, w := 0, 0; i < len(quotedString); i += w {
+		c, width := utf8.DecodeRuneInString(quotedString[i:])
+		read += width
+		w = width
+
+		switch c {
+		case '"':
+			if escaping {
+				result.WriteRune(c)
+				escaping = false
+			} else {
+				closedQuote = true
+				break
+			}
+		case '\\':
+			if escaping {
+				result.WriteRune(c)
+				escaping = false
+			} else {
+				escaping = true
+			}
+		default:
+			result.WriteRune(c)
+			escaping = false
+		}
+	}
+
+	if !closedQuote {
+		return "", 0, errors.New("invalid quoted string: missing closing quote")
+	}
+	return result.String(), read, nil
+}
+
+func NewWarningHeader(code int, agent, text string) (string, error) {
```

Comment:
Could we structure the warning with json rather than having our own format, and then escape things etc?

Reply:
Should it be an API type? Should we consider versioning it?